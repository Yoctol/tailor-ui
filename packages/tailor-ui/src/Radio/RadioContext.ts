import { createContext } from 'react';

export type Direction = 'horizontal' | 'vertical';

const RadioContext = createContext<{
  _onChange?: (value: string) => void;
  _isChecked?: (value: string) => boolean;
  direction: Direction;
}>({
  direction: 'horizontal',
});

export { RadioContext };
