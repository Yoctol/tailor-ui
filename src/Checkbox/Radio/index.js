import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { themeGet } from 'styled-system';
import { ifProp, switchProp } from 'styled-tools';

import theme from '../../theme';

const getCheckmarkSpace = size => css`
  width: calc(${themeGet(size)} * 1.5);
  height: calc(${themeGet(size)} * 1.5);
`;

const Checkmark = styled.span`
  height: 0;
  margin-right: 0.25rem;
  border: 1px solid
    ${ifProp('disabled', themeGet('colors.border'), themeGet('colors.primary'))};
  border-radius: 50%;
  background-color: #fff;
  cursor: ${ifProp('disabled', 'default', 'pointer')};

  &::after {
    content: '';
    display: none;
    position: absolute;
    border-radius: 50%;
    background: ${themeGet('colors.primary')};
    transform: translate(-50%, -50%);

    ${switchProp('size', {
      sm: getCheckmarkSpace('space.paddingXSm'),
      m: getCheckmarkSpace('space.paddingX'),
      lg: getCheckmarkSpace('space.paddingXLg'),
    })};
  }

  ${switchProp('size', {
    sm: css`
      padding: ${themeGet('space.paddingXSm')} ${themeGet('space.paddingXSm')};
    `,
    m: css`
      padding: ${themeGet('space.paddingX')} ${themeGet('space.paddingX')};
    `,
    lg: css`
      padding: ${themeGet('space.paddingXLg')} ${themeGet('space.paddingXLg')};
    `,
  })};
`;

Checkmark.propTypes = {
  size: PropTypes.oneOf(['sm', 'm', 'lg']).isRequired,
};

Checkmark.defaultProps = {
  theme,
};

const Container = styled.div`
  display: inline-flex;
  align-items: center;
  user-select: none;

  input {
    position: absolute;
    opacity: 0;

    &:checked {
      /* stylelint-disable-next-line */
      & ~ ${Checkmark}::after {
        display: block;
      }
    }
  }
`;

const Radio = ({ disabled, size, ...props }) => (
  <Container>
    <input type="radio" disabled={disabled} {...props} />
    <Checkmark size={size} disabled={disabled} />
  </Container>
);

Radio.propTypes = {
  disabled: PropTypes.bool,
  name: PropTypes.string,
  size: PropTypes.oneOf(['sm', 'm', 'lg']),
};

Radio.defaultProps = {
  disabled: false,
  name: 'radio',
  size: 'm',
};

export default Radio;
