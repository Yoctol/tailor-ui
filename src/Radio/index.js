import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import { borderColor, color, themeGet } from 'styled-system';

import Label from '../Form/Label';
import { controlShadow } from '../utils/shadow';

const StyledRadio = styled.input.attrs({
  type: 'radio',
})`
  position: absolute;
  opacity: 0;

  & + ${Label /* sc-selector */} {
    position: relative;
    padding: 0;
    color: ${themeGet('colors.gray.3')};
    cursor: pointer;
  }

  & + ${Label /* sc-selector */}::before {
    content: '';
    display: inline-block;
    width: 16px;
    height: 16px;
    margin-right: ${themeGet('space.1')};
    border: ${themeGet('borders.default')};
    border-radius: 999px;
    background: white;
    vertical-align: text-bottom;
    ${borderColor};
  }

  & + ${Label /* sc-selector */}::after {
    content: '';
    position: absolute;
    top: 2px;
    left: 2px;
    width: 12px;
    height: 12px;
    border-radius: 50%;
    ${p => p.theme.transition /* sc-declaration */};
  }

  &:focus + ${Label /* sc-selector */}::before {
    ${controlShadow()};
  }

  &:disabled + ${Label /* sc-selector */} {
    opacity: 0.5;
    cursor: auto;
    pointer-events: none;
  }

  &:disabled + ${Label /* sc-selector */}::before {
    opacity: 0.5;
  }

  &:checked + ${Label /* sc-selector */}::after {
    ${color};
  }
`;

const Radio = props => <StyledRadio {...props} />;

Radio.propTypes = {
  /**
   * Specifies whether the checkbox is selected
   */
  checked: PropTypes.bool,
  /**
   * Disable checkbox
   */
  disabled: PropTypes.bool,
  /**
   * The callback function that is triggered when the state changes
   */
  onChange: PropTypes.func,
  ...borderColor.propTypes,
  ...color.propTypes,
};

Radio.defaultProps = {
  disabled: false,
  checked: false,
  onChange: () => {},
  // eslint-disable-next-line react/default-props-match-prop-types
  bg: 'primary',
  // eslint-disable-next-line react/default-props-match-prop-types
  borderColor: 'primary',
};

export default Radio;
