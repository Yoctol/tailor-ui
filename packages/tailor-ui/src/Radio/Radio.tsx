import React, {
  ChangeEventHandler,
  FocusEventHandler,
  MouseEventHandler,
  ReactNode,
  forwardRef,
  useContext,
} from 'react';

import { Box } from '../Layout';
import { useFormField } from '../FormField';

import { RadioContext } from './RadioContext';
import { RadioInner, RadioLabel, RadioWrapper, StyledRadio } from './styles';

export interface RadioProps {
  children?: ReactNode;
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  value?: string;
  onFocus?: FocusEventHandler<HTMLInputElement>;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onMouseEnter?: MouseEventHandler<HTMLLabelElement>;
  onMouseLeave?: MouseEventHandler<HTMLLabelElement>;
}

const Radio = forwardRef<HTMLLabelElement, RadioProps>(function Radio(
  {
    children,
    disabled = false,
    checked,
    defaultChecked,
    onChange,
    value,
    onMouseEnter,
    onMouseLeave,
    ...props
  },
  ref
) {
  const { handleChange, isChecked, direction } = useContext(RadioContext);
  const [, , setValue] = useFormField({
    value: checked,
    defaultValue: defaultChecked,
  });

  const inGroup = Boolean(handleChange);
  const boxChecked = isChecked && value ? isChecked(value) : checked;

  return (
    <RadioLabel
      ref={ref}
      disabled={disabled}
      direction={direction}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      <RadioWrapper>
        <StyledRadio
          disabled={disabled}
          checked={boxChecked}
          defaultChecked={defaultChecked}
          onChange={event => {
            if (onChange) {
              onChange(event);
            }
            if (handleChange && value) {
              handleChange(value);
            }
            if (!inGroup) {
              setValue(event.target.checked);
            }
          }}
          {...props}
        />
        <RadioInner />
      </RadioWrapper>
      <Box px="2">{children}</Box>
    </RadioLabel>
  );
});

export { Radio };
