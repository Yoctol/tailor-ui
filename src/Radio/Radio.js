import PropTypes from 'prop-types';
import React from 'react';
import styled, { css } from 'styled-components';

import Space from '../Grid/Space';

import { Consumer } from './RadioContext';

const RadioWrapper = styled.span`
  display: inline-block;
  position: relative;
  margin-bottom: 1px;
  line-height: 1;
`;

const RadioInner = styled.span`
  display: block;
  position: relative;
  width: 16px;
  height: 16px;
  padding: 0;
  border: ${p => p.theme.borders.base};
  border-radius: 50%;
  border-color: ${p => p.theme.colors.gray[7]};
  background-color: ${p => p.theme.colors.light};
  color: ${p => p.theme.colors.gray[3]};

  ${p => p.theme.transition};

  &::after {
    content: '';
    position: absolute;
    top: 2px;
    left: 2px;
    width: 10px;
    height: 10px;
    border-radius: 50%;
    opacity: 0;
    background-color: ${p => p.theme.colors.primary};
    transform: scale(0);
    ${p => p.theme.transition};
  }
`;

const StyledRadio = styled.input.attrs({
  type: 'radio',
})`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;

  &:checked + ${RadioInner /* sc-selector */} {
    border-color: ${p => p.theme.colors.primary};
  }

  &:checked + ${RadioInner /* sc-selector */}::after {
    opacity: 1;
    transform: scale(1);
  }
`;

const getMarginPosition = p =>
  p.direction === 'horizontal' ? 'margin-left' : 'margin-top';

const RadioLabelBase = styled.label`
  display: inline-flex;
  align-items: center;
  font-size: ${p => p.theme.fontSizes.base};
  cursor: ${p => (p.disabled ? 'not-allowed' : 'pointer')};

  ${p =>
    p.disabled
      ? css`
          opacity: 0.5;
        `
      : css`
          &:hover ${RadioInner /* sc-selector */} {
            border-color: ${p.theme.colors.primary};
          }
        `};
`;

const RadioLabel = styled(RadioLabelBase)`
  /* stylelint-disable-next-line no-descending-specificity */
  & + ${RadioLabelBase /* sc-selector */} {
    ${getMarginPosition /* sc-prop */}: ${p => p.theme.space[2]};
  }
`;

const Radio = ({
  children,
  disabled,
  checked,
  defaultChecked,
  onChange,
  value,
  ...props
}) => (
  <Consumer>
    {({ _onChange, _isChecked, direction }) => (
      <RadioLabel disabled={disabled} direction={direction} {...props}>
        <RadioWrapper>
          <StyledRadio
            disabled={disabled}
            checked={_isChecked ? _isChecked(value) : checked}
            defaultChecked={defaultChecked}
            onChange={event => {
              onChange(event);
              if (_onChange) _onChange(value);
            }}
          />
          <RadioInner />
        </RadioWrapper>
        <Space px="2">{children}</Space>
      </RadioLabel>
    )}
  </Consumer>
);

Radio.propTypes = {
  /**
   * Specifies whether the Radio is selected
   */
  // eslint-disable-next-line react/require-default-props
  checked: PropTypes.bool,
  /**
   * Specifies the initial state: whether or not the Radio is selected.
   */
  // eslint-disable-next-line react/require-default-props
  defaultChecked: PropTypes.bool,
  /**
   * Disable Radio
   */
  disabled: PropTypes.bool,
  /**
   * The callback function that is triggered when the state changes
   */
  onChange: PropTypes.func,
};

Radio.defaultProps = {
  disabled: false,
  onChange: () => {},
};

export default Radio;
