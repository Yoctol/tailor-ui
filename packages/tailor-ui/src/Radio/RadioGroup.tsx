import React, { FunctionComponent, useState } from 'react';
import styled from 'styled-components';

import Radio from './Radio';
import RadioContext, { Direction } from './RadioContext';

const RadioGroupFlex = styled.div<{ direction: Direction }>`
  display: ${p => (p.direction === 'horizontal' ? 'flex' : 'inline-flex')};
  flex-direction: ${p => (p.direction === 'horizontal' ? 'row' : 'column')};
`;

export interface RadioGroupProps {
  /**
   * To set the initial value
   */
  defaultValue?: string;
  /**
   * to specify the direction of the step bar, `horizontal` and `vertical` are currently supported
   */
  direction?: Direction;
  /**
   * Specifies options
   */
  options?: {
    label: string;
    value: string;
    disabled?: boolean;
  }[];
  /**
   * Used for setting the currently selected value
   */
  value?: string;
  /**
   * The callback function that is triggered when the state changes
   */
  onChange?: (value: string) => void;
}

const RadioGroup: FunctionComponent<RadioGroupProps> = ({
  value: controlledValue,
  defaultValue,
  options = null,
  onChange,
  direction = 'horizontal' as Direction,
  children,
  ...otherProps
}) => {
  const [uncontrolledValue, setUncontrolledValue] = useState(defaultValue);

  return (
    <RadioContext.Provider
      value={{
        direction,
        _onChange: _value => {
          if (uncontrolledValue) {
            setUncontrolledValue(_value);
          }

          if (onChange) {
            onChange(_value);
          }
        },
        _isChecked: _value =>
          controlledValue
            ? controlledValue === _value
            : uncontrolledValue === _value,
      }}
    >
      <RadioGroupFlex direction={direction}>
        {options
          ? options.map(({ label, value: optionValue, disabled = false }) => (
              <Radio
                key={label}
                value={optionValue}
                disabled={disabled}
                {...otherProps}
              >
                {label}
              </Radio>
            ))
          : children}
      </RadioGroupFlex>
    </RadioContext.Provider>
  );
};

export default RadioGroup;
