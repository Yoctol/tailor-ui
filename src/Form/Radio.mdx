---
name: Radio
menu: Data Entry
route: /data-entry/radio
---

import { PropsTable, Playground } from 'docz';
import { Toggle, Value } from 'react-powerplug';
import FormField from './FormField';
import Label from './Label';
import Radio from '../Radio';
import RadioField, { RadioFieldGroup } from './RadioField';


# Radio

## Properties

### Radio

<PropsTable of={Radio} />

### RadioField

<PropsTable of={RadioField} />

### RadioField.Group

<PropsTable of={RadioFieldGroup} />

## Usage

```js
import { FormField, Label, Radio, RadioField } from '@yoctol/ui';
```

### Radio

#### Basic

<Playground>
  <FormField>
    <Radio name="radio-basic" id="radio-basic-a" defaultChecked />
    <Label htmlFor="radio-basic-a">Radio A</Label>
  </FormField>
  <FormField>
    <Radio name="radio-basic" id="radio-basic-b" />
    <Label htmlFor="radio-basic-b">Radio B</Label>
  </FormField>
</Playground>

#### With disabled

<Playground>
  <FormField>
    <Radio
      name="radio-disabled"
      id="radio-disable-a"
      checked
      disabled
    />
    <Label htmlFor="radio-disable-a">Checked Disabled</Label>
  </FormField>
  <FormField>
    <Radio name="radio-disabled" id="radio-disable-b" disabled />
    <Label htmlFor="radio-disable-b">Disabled</Label>
  </FormField>
</Playground>

#### Custom color

<Playground>
  <FormField>
    <Radio
      name="customized"
      id="custom-a"
      bg="#63bf2d"
      borderColor="#423b63"
      defaultChecked
    />
    <Label htmlFor="custom-a">Customized Color Radio A</Label>
  </FormField>
  <FormField>
    <Radio
      name="customized"
      id="custom-b"
      bg="#63bf2d"
      borderColor="#423b63"
    />
    <Label htmlFor="custom-b">Customized Color Radio B</Label>
  </FormField>
</Playground>


### RadioField

#### Basic

<Playground>
  <Value initial="radio-1">
    {({ value, set }) => (
      <div>
        <RadioField checked={value === 'radio-1'} onChange={() => set('radio-1')}>
          Radio
        </RadioField>
        <RadioField checked={value === 'radio-2'} onChange={() => set('radio-2')}>
          Radio
        </RadioField>
      </div>
    )}
  </Value>
</Playground>

### RadioField.Group

#### Basic

<Playground>
  <Value initial="radio-a">
    {({ set, value }) => (
      <RadioField.Group
        label="Radio Group"
        options={[
          { label: 'Radio A', value: 'radio-a' },
          { label: 'Radio B', value: 'radio-b' },
          { label: 'Radio C', value: 'radio-c', disabled: true },
          { label: 'Radio D', value: 'radio-d' },
        ]}
        value={value}
        onChange={set}
      />
    )}
  </Value>
</Playground>

#### With group & error

<Playground>
  <Value initial="radio-a">
    {({ set, value }) => (
      <RadioField.Group
        label="Radio Group"
        error
        message="Something went wrong!"
        options={[
          { label: 'Radio A', value: 'radio-a' },
          { label: 'Radio B', value: 'radio-b' },
          { label: 'Radio C', value: 'radio-c', disabled: true },
          { label: 'Radio D', value: 'radio-d' },
        ]}
        value={value}
        onChange={set}
      />
    )}
  </Value>
</Playground>
