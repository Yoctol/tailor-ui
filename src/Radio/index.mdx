---
name: Radio
menu: Components
route: /components/radio
---

import { PropsTable, Playground } from 'docz';

import Button from '../Button';
import Box from '../Grid/Box';

import Radio from './';
import RadioGroup from './RadioGroup';

# Radio

Radio.

## When To Use

- Used to select a single state in multiple options.
- The difference between Select is that Radio is visible to user and can facilitate the comparison of choice, which makes there shouldn't be too many of them.

## Examples

```js
import { Radio } from 'tailor-ui';
```

### Radio

#### Basic

<Playground>
  <Radio>Radio</Radio>
  <br />
  <Radio defaultChecked>Radio</Radio>
</Playground>

#### Disabled

<Playground>
  <Radio checked disabled>
    Radio
  </Radio>
  <br />
  <Radio disabled>Disabled</Radio>
</Playground>

### Radio.Group

#### Basic

<Playground>
  {() => {
    const [value, setValue] = React.useState('radio-a');

    return (
      <>
        <Radio.Group
          options={[
            { label: 'Radio A', value: 'radio-a' },
            { label: 'Radio B', value: 'radio-b' },
            { label: 'Radio C', value: 'radio-c', disabled: true },
            { label: 'Radio D', value: 'radio-d' },
          ]}
          value={value}
          onChange={setValue}
        />
        <br />
        <Radio.Group
          options={[
            { label: 'Radio A', value: 'radio-a' },
            { label: 'Radio B', value: 'radio-b' },
            { label: 'Radio C', value: 'radio-c', disabled: true },
            { label: 'Radio D', value: 'radio-d' },
          ]}
          defaultValue="radio-b"
          onChange={console.log}
        />
      </>
    );

}}

</Playground>

#### Controlled

<Playground>
  {() => {
    const [value, setValue] = React.useState('radio-a');

    return (
      <>
        <Radio.Group
          options={[
            { label: 'Radio A', value: 'radio-a' },
            { label: 'Radio B', value: 'radio-b' },
            { label: 'Radio C', value: 'radio-c', disabled: true },
            { label: 'Radio D', value: 'radio-d' },
          ]}
          value={value}
          onChange={setValue}
        />
        <br />
        <Button onClick={() => set('radio-d')}>Check last one</Button>
      </>
    );

}}

</Playground>

#### Composition render

<Playground>
  {() => {
    const [value, setValue] = React.useState('radio_1');

    return (
      <>
        <Radio.Group value={value} onChange={setValue}>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_1">radio 1</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_2">radio 2</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_3">radio 3</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_4">radio 4</Radio>
          </Box>
        </Radio.Group>
        <br />
        <Radio.Group defaultValue="radio_1" onChange={console.log}>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_1">radio 1</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_2">radio 2</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_3">radio 3</Radio>
          </Box>
          <Box bg="gray300" p="2" m="2">
            <Radio value="radio_4">radio 4</Radio>
          </Box>
        </Radio.Group>
      </>
    );

}}

</Playground>

#### Direction

<Playground>
  {() => {
    const [value, setValue] = React.useState('');

    return (
      <Radio.Group
        direction="vertical"
        options={[
          { label: 'radio 1', value: 'radio_1' },
          { label: 'radio 2', value: 'radio_2' },
          { label: 'radio 3', value: 'radio_3' },
          { label: 'radio 4', value: 'radio_4' },
        ]}
        value={value}
        onChange={setValue}
      />
    );

}}

</Playground>

## API

### Radio

<PropsTable of={Radio} />

### Radio.Group

<PropsTable of={RadioGroup} />
