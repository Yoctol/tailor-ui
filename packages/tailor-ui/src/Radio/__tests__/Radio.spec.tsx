import React from 'react';

import { fireEvent, render } from 'test/test-utils';

import { Radio } from '../Radio';

describe('Radio', () => {
  it('should render correctly', () => {
    const { container } = render(<Radio>Radio</Radio>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render disabled radio', () => {
    const { container } = render(<Radio disabled>Radio</Radio>);

    const radio = container.querySelector('input[type=radio]');

    expect(radio).toBeDisabled();
  });

  it('should render checked radio', () => {
    const { container } = render(<Radio checked>Radio</Radio>);

    const radio = container.querySelector('input[type=radio]');

    expect(radio).toHaveAttribute('checked');
  });

  it('should call onChange when radio is clicked', () => {
    const onChange = jest.fn();

    const { container } = render(
      <Radio onChange={onChange} checked={false}>
        Radio
      </Radio>
    );

    const radio = container.querySelector(
      'input[type=radio]'
    ) as HTMLInputElement;

    fireEvent.click(radio);

    expect(onChange).toBeCalled();
  });

  it('should support data-testid', () => {
    const { getByTestId } = render(
      <Radio checked data-testid="my-radio">
        Radio
      </Radio>
    );

    expect(getByTestId('my-radio')).toHaveAttribute('checked');
  });
});
