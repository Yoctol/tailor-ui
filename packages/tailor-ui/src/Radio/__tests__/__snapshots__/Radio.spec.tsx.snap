// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should render correctly 1`] = `
.c6 {
  box-sizing: border-box;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
}

.c2 {
  display: inline-block;
  position: relative;
  margin-bottom: 1px;
  line-height: 1;
}

.c5 {
  display: block;
  position: relative;
  width: 16px;
  height: 16px;
  padding: 0;
  border: 1px solid;
  border-radius: 50%;
  border-color: #bfbfbf;
  background-color: #ffffff;
  color: #373737;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c5::after {
  content: '';
  position: absolute;
  top: 2px;
  left: 2px;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  opacity: 0;
  background-color: #3c5ad0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c3 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
}

.c3:checked + .c4 {
  border-color: #3c5ad0;
}

.c3:checked + .c4::after {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1rem;
  cursor: pointer;
}

.c1:hover .c4 {
  border-color: #3c5ad0;
}

.c1 + .c0 {
  margin-left: 0.5rem;
}

<label
  class="c0 c1"
  direction="horizontal"
>
  <span
    class="c2"
  >
    <input
      class="c3"
      type="radio"
    />
    <span
      class="c4 c5"
    />
  </span>
  <div
    class="c6"
  >
    Radio
  </div>
</label>
`;
